source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,412 @@
+          allowed_compressors.begin(), allowed_compressors.end(),
+          [](const std::pair<std::string, uint32_t>& a,
+             const std::pair<std::string, uint32_t>& b) -> bool { return a.second < b.second; });
+      return std::make_unique<CompressorFilter::EncodingDecision>(
+          first_registered->first, CompressorFilter::EncodingDecision::HeaderStat::Wildcard);
+    }
+  }
+
+  if (absl::EqualsIgnoreCase(config_->contentEncoding(), choice.first)) {
+    config_->stats().header_compressor_used_.inc();
+    // TODO(rojkov): Remove this increment when the gzip-specific stat is gone.
+    if (absl::EqualsIgnoreCase("gzip", choice.first)) {
+      config_->stats().header_gzip_.inc();
+    }
+    return std::make_unique<CompressorFilter::EncodingDecision>(
+        std::string(choice.first), CompressorFilter::EncodingDecision::HeaderStat::Used);
+  } else if (!allowed_compressors.empty()) {
+    config_->stats().header_compressor_overshadowed_.inc();
+    return std::make_unique<CompressorFilter::EncodingDecision>(
+        std::string(choice.first), CompressorFilter::EncodingDecision::HeaderStat::Overshadowed);
+  }
+
+  config_->stats().header_not_valid_.inc();
+  return std::make_unique<CompressorFilter::EncodingDecision>(
+      Http::Headers::get().AcceptEncodingValues.Identity,
+      CompressorFilter::EncodingDecision::HeaderStat::NotValid);
+}
+
+bool CompressorFilter::isAcceptEncodingAllowed(const Http::ResponseHeaderMap& headers) const {
+  if (accept_encoding_ == nullptr) {
+    config_->stats().no_accept_header_.inc();
+    return false;
+  }
+
+  const absl::string_view encoding_decision_key{"encoding_decision"};
+
+  // Check if we have already cached our decision on encoding.
+  const StreamInfo::FilterStateSharedPtr& filter_state =
+      decoder_callbacks_->streamInfo().filterState();
+  if (filter_state->hasData<CompressorFilter::EncodingDecision>(encoding_decision_key)) {
+    const CompressorFilter::EncodingDecision& decision =
+        filter_state->getDataReadOnly<CompressorFilter::EncodingDecision>(encoding_decision_key);
+    if (absl::EqualsIgnoreCase(config_->contentEncoding(), decision.encoding())) {
+      config_->stats().header_compressor_used_.inc();
+      // TODO(rojkov): Remove this increment when the gzip-specific stat is gone.
+      if (absl::EqualsIgnoreCase("gzip", config_->contentEncoding())) {
+        config_->stats().header_gzip_.inc();
+      }
+      return true;
+    }
+
+    switch (decision.stat()) {
+    case CompressorFilter::EncodingDecision::HeaderStat::Identity:
+      config_->stats().header_identity_.inc();
+      break;
+    case CompressorFilter::EncodingDecision::HeaderStat::Wildcard:
+      config_->stats().header_wildcard_.inc();
+      break;
+    case CompressorFilter::EncodingDecision::HeaderStat::NotValid:
+      config_->stats().header_not_valid_.inc();
+      break;
+    default:
+      config_->stats().header_compressor_overshadowed_.inc();
```

Comment:
What does overshadowed mean? Note that this will need documentation once the filter itself is actually created.

Reply:
Yes, the refactored gzip filter includes this stat documented in its .rst file. For now I added the description to `compressor.h`.